src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2055,458 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+		}
+
+		/*
+		 * We are only dealing with hash distributed tables, that's why we
+		 * can hard code data type and typemod.
+		 */
+		const int intervalTypeId = INT4OID;
+		const int intervalTypeMod = -1;
+
+		Relation distShardRelation = table_open(DistShardRelationId(), AccessShareLock);
+		TupleDesc distShardTupleDesc = RelationGetDescr(distShardRelation);
+
+		FmgrInfo *shardIntervalCompareFunction =
+			GetFunctionInfo(intervalTypeId, BTREE_AM_OID, BTORDER_PROC);
+
+		HeapTuple shardTuple = NULL;
+		foreach_ptr(shardTuple, distShardTupleList)
+		{
+			ShardInterval *shardInterval =
+				TupleToShardInterval(shardTuple, distShardTupleDesc,
+									 intervalTypeId, intervalTypeMod);
+
+			Datum firstMin = Int32GetDatum(shardMinValueInt);
+			Datum firstMax = Int32GetDatum(shardMaxValueInt);
+			Datum secondMin = shardInterval->minValue;
+			Datum secondMax = shardInterval->maxValue;
+			Oid collationId = InvalidOid;
+
+			if (ShardIntervalsOverlapWithParams(firstMin, firstMax, secondMin, secondMax,
+												shardIntervalCompareFunction,
+												collationId))
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("Shard intervals overlap for table \"%s\": "
+									   "%ld and %ld", get_rel_name(relationId),
+									   shardId, shardInterval->shardId)));
+			}
+		}
+
+		table_close(distShardRelation, NoLock);
+	}
+}
+
+
+/*
+ * citus_internal_add_placement_metadata is an internal UDF to
+ * add a row to pg_dist_placement.
+ */
+Datum
+citus_internal_add_placement_metadata(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	int64 shardId = PG_GETARG_INT64(1);
+	int32 shardState = PG_GETARG_INT32(2);
+	int64 shardLength = PG_GETARG_INT64(3);
+	int32 groupId = PG_GETARG_INT32(4);
+	int64 placementId = PG_GETARG_INT64(5);
+
+	/* only owner of the table is allowed to modify the metadata */
+	EnsureTableOwner(relationId);
+
+	/* we want to serialize all the metadata changes to this table */
+	LockRelationOid(relationId, ExclusiveLock);
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/*
+		 * Even if the table owner is a malicious user, as long as the shard placements
+		 * fit into basic requirements of Citus metadata, the user can only affect its
+		 * own tables. Given that the user is owner of the table, we should allow.
+		 */
+		EnsureShardPlacementMetadataIsSane(shardId, placementId, shardState,
+										   shardLength, groupId);
+	}
+
+	InsertShardPlacementRow(shardId, placementId, shardState, shardLength, groupId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureShardPlacementMetadataIsSane ensures if the input parameters for
+ * the shard placement metadata is sane.
+ */
+static void
+EnsureShardPlacementMetadataIsSane(int64 shardId, int64 placementId,
+								   int32 shardState, int64 shardLength,
+								   int32 groupId)
+{
+	if (shardId <= INVALID_SHARD_ID)
```

Comment:
I think we should also check that the `shardId` actually exists in `pg_dist_shard`.

Reply:
makes sense